internal/manager/setup.go

```diff
@@ -192,51 +192,88 @@ func setupAdmissionServer(
+// set or the publish service addresses if no overrides are set. If no UDP overrides or UDP publish service are set,
+// the UDP finder will also return the default addresses. If no override or publish service is set, this function
+// returns nil finders and an error.
+func setupDataplaneAddressFinder(
+	ctx context.Context,
+	mgrc client.Client,
+	c *Config,
+) (*dataplane.AddressFinder, *dataplane.AddressFinder, error) {
 	dataplaneAddressFinder := dataplane.NewAddressFinder()
+	udpDataplaneAddressFinder := dataplane.NewAddressFinder()
+	var getter func() ([]string, error)
 	if c.UpdateStatus {
+		// Default
 		if overrideAddrs := c.PublishStatusAddress; len(overrideAddrs) > 0 {
 			dataplaneAddressFinder.SetOverrides(overrideAddrs)
 		} else if c.PublishService != "" {
 			parts := strings.Split(c.PublishService, "/")
 			if len(parts) != 2 {
-				return nil, fmt.Errorf("publish service %s is invalid, expecting <namespace>/<name>", c.PublishService)
+				return nil, nil, fmt.Errorf("publish service %s is invalid, expecting <namespace>/<name>", c.PublishService)
 			}
 			nsn := types.NamespacedName{
 				Namespace: parts[0],
 				Name:      parts[1],
 			}
-			dataplaneAddressFinder.SetGetter(func() ([]string, error) {
-				svc := new(corev1.Service)
-				if err := mgrc.Get(ctx, nsn, svc); err != nil {
-					return nil, err
-				}
+			getter = generateAddressFinderGetter(ctx, mgrc, nsn)
+			dataplaneAddressFinder.SetGetter(getter)
+		} else {
+			return nil, nil, fmt.Errorf("status updates enabled but no method to determine data-plane addresses, need either --publish-service or --publish-status-address")
+		}
 
-				var addrs []string
-				switch svc.Spec.Type { //nolint:exhaustive
-				case corev1.ServiceTypeLoadBalancer:
-					for _, lbaddr := range svc.Status.LoadBalancer.Ingress {
-						if lbaddr.IP != "" {
-							addrs = append(addrs, lbaddr.IP)
-						}
-						if lbaddr.Hostname != "" {
-							addrs = append(addrs, lbaddr.Hostname)
-						}
-					}
-				default:
-					addrs = append(addrs, svc.Spec.ClusterIPs...)
-				}
+		// UDP. falls back to default if not configured
+		if udpOverrideAddrs := c.PublishStatusAddressUDP; len(udpOverrideAddrs) > 0 {
+			udpDataplaneAddressFinder.SetUDPOverrides(udpOverrideAddrs)
+		} else if c.PublishServiceUDP != "" {
+			parts := strings.Split(c.PublishServiceUDP, "/")
+			if len(parts) != 2 {
+				return nil, nil, fmt.Errorf("UDP publish service %s is invalid, expecting <namespace>/<name>", c.PublishService)
+			}
+			nsn := types.NamespacedName{
+				Namespace: parts[0],
+				Name:      parts[1],
+			}
+			udpDataplaneAddressFinder.SetGetter(generateAddressFinderGetter(ctx, mgrc, nsn))
+		} else {
+			udpDataplaneAddressFinder.SetGetter(getter)
+		}
+	}
+
+	return dataplaneAddressFinder, udpDataplaneAddressFinder, nil
+}
+
+func generateAddressFinderGetter(
+	ctx context.Context,
+	mgrc client.Client,
+	nsn types.NamespacedName,
+) func() ([]string, error) {
```

Comment:
This should be returning a func that accepts a context. But as noted above: we can address this separately.